File: android/app/src/main/java/com/example/paperclipper/UserId.kt
package com.example.paperclipper

import android.content.Context
import com.google.firebase.auth.FirebaseAuth
import java.util.UUID

/**
 * Stable identifier for the per-user daily quota the server enforces. Uses the signed-in Firebase
 * UID when available, otherwise a persistent per-install UUID kept in SharedPreferences.
 */
object UserId {
    fun get(context: Context): String {
        val uid = runCatching { FirebaseAuth.getInstance().currentUser?.uid }.getOrNull()
        if (uid != null) return "uid:$uid"

        val prefs = context.getSharedPreferences("paperclipper", Context.MODE_PRIVATE)
        prefs.getString("install_id", null)?.let { return "dev:$it" }
        val fresh = UUID.randomUUID().toString()
        prefs.edit().putString("install_id", fresh).apply()
        return "dev:$fresh"
    }
}

File: ios/PaperClipper/Networking/UserId.swift
import Foundation

#if canImport(FirebaseCore)
import FirebaseCore
import FirebaseAuth
#endif

/// Stable identifier for the per-user daily quota the server enforces (sent as the `X-User-Id`
/// header on `/analyze`). Uses the signed-in Firebase UID when available, otherwise a persistent
/// per-install UUID kept in `UserDefaults`. Mirrors Android's `UserId.kt` (`uid:` vs `dev:` prefix).
enum UserId {
    private static let installIdKey = "paperclipper.install_id"

    static func get() -> String {
        #if canImport(FirebaseCore)
        // Guard on FirebaseApp being configured — touching Auth before configure() crashes.
        if FirebaseApp.app() != nil, let uid = Auth.auth().currentUser?.uid {
            return "uid:\(uid)"
        }
        #endif
        let defaults = UserDefaults.standard
        if let existing = defaults.string(forKey: installIdKey) {
            return "dev:\(existing)"
        }
        let fresh = UUID().uuidString
        defaults.set(fresh, forKey: installIdKey)
        return "dev:\(fresh)"
    }
}
